### PR TITLE
refactor(swap): enhance balance checker and protocol fee tests

### PR DESCRIPTION
**Pull request by jinoosss**
### Background
In `ExactOut` swaps, router fee handling can make it hard to reason about user-received output, pool output, and protocol fee consistency.  
This PR improves error observability and expands test coverage to validate exact-out fee accounting across realistic routing scenarios.

### What Changed
- Added a new router error:
  - `GNOSWAP-ROUTER-018`: insufficient balance for swap
- Improved `SwapCallback` balance checks:
  - In the `transferFrom` path, the router now validates payer balance and panics with detailed context (`token`, `required`, `available`, `payer`) when insufficient.
- Updated existing unit test behavior:
  - Switched from exact panic message matching to partial matching (`AbortsContains`) in `swap_inner_test`.
- Extended exact-out fee unit tests in `protocol_fee_swap_test`:
  - Round-trip correctness
  - Fee accounting consistency
  - Small-amount rounding boundary cases
- Added 4 scenario filetests:
  - `exact_out_router_fee_verification_filetest.gno`
  - `exact_out_router_fee_edge_cases_filetest.gno`
  - `exact_out_amount_verification_filetest.gno`
  - `exact_out_dryswap_fee_detection_filetest.gno`

### Why
- Ensure `ExactOut` keeps the core invariant: user receives exactly requested `amountOut`.
- Confirm accounting invariant: `protocol fee = pool output - user output`.
- Reduce regression risk across single-hop, multi-hop, split routes, fee tiers, and rounding boundaries.
- Provide clearer failure diagnostics for insufficient-balance callback paths.

### 1) ExactOut Router Fee Verification
File: `contract/r/scenario/router/exact_out_router_fee_verification_filetest.gno`

- Setup 1:1 pools (`bar:baz`, `baz:foo`) with deep liquidity.
- Run single-hop exact-out at 0.15% router fee.
- Run single-hop exact-out at 1% router fee.
- Run multi-hop exact-out at 0.15% router fee.
- Run multi-hop exact-out at 1% router fee.
- Verify:
  - User receives exactly the requested output amount.
  - Protocol fee matches `pool output - user output`.

### 2) ExactOut Router Fee Edge Cases
File: `contract/r/scenario/router/exact_out_router_fee_edge_cases_filetest.gno`

- Setup deep-liquidity single pool.
- Execute exact-out with minimum amount (`1` token).
- Execute exact-out with large amount (`9,000,000,000,000` tokens).
- Verify:
  - Correct truncation behavior on tiny amounts (fee can truncate to `0`).
  - Accounting and output invariants remain correct for large amounts.

### 3) ExactOut Amount Verification (Comprehensive)
File: `contract/r/scenario/router/exact_out_amount_verification_filetest.gno`

- Setup 4 pools including mixed fee tiers (`3000`, `500`).
- Run:
  - 1-route 1-hop
  - 1-route 2-hop
  - 1-route 3-hop
  - 2-route split (50/50)
  - minimum amount
  - large amount
  - low-fee pool route
- Verify:
  - Exact requested output is received in each case.
  - Fee accounting consistency where applicable.

### 4) Fee Detection for ExactOut
File: `contract/r/scenario/router/exact_out_dryswap_fee_detection_filetest.gno`

- Initialize user balances and pool state.
- Compare `DrySwapRoute` quotes at fee `0` vs fee `15`.
- Compare quote behavior against actual exact-out execution.
- Verify:
  - Dry-swap quote reflects increased required input when router fee is applied.
  - Quote behavior is consistent with execution expectations under fee overhead.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved insufficient-balance handling for swaps: errors are now tagged and include detailed context (token, required vs available amounts, payer), and tests accept the new tagged message format.

* **Tests**
  * Added comprehensive unit tests for fee calculation (edge cases, large values, and rounding) and updated callback tests to assert the new error behavior.

### CODE REVIEW

File: contract/r/gnoswap/router/v1/swap_callback.gno [hunk 1]
```diff
@@ -68,18 +69,22 @@ func (r *routerV1) SwapCallback(
...
-	}
+	amountInt64 := amount.Int64()
 
 	poolAddr := access.MustGetAddress(prbac.ROLE_POOL.String())
 	routerAddr := access.MustGetAddress(prbac.ROLE_ROUTER.String())
 
 	if payer == routerAddr {
-		common.SafeGRC20Transfer(cross, token, poolAddr, amount.Int64())
+		common.SafeGRC20Transfer(cross, token, poolAddr, amountInt64)
 	} else {
```
Comment (junghoon-vans) [inline contract/r/gnoswap/router/v1/swap_callback.gno:86]: Was there a reason to narrow the explicit insufficient-balance check to only the `transferFrom` path? Previously this guard applied regardless of payer, so router-funded hops now seem to fall back to `SafeGRC20Transfer` failure semantics instead of returning `GNOSWAP-ROUTER-018`. If that distinction is intentional, it would be good to document why.

**Review by junghoon-vans — APPROVED** [review]
